### PR TITLE
Generalise preprocessing

### PR DESCRIPTION
**Pull request by PJ-Prins**

### CODE REVIEW

File: src/private/fnft__nse_discretization.c [hunk 1]
```diff
@@ -383,274 +383,29 @@ INT fnft__nse_discretization_phase_factor_b(const REAL eps_t, const UINT D, REAL
...
-            }
-            
-            REAL eps_t_sub = eps_t*nskip_per_step;
-            REAL eps_t_sub_2 = POW(eps_t_sub,2);        
-            q_preprocessed[1] = (q_preprocessed[3]-0)/(2*eps_t_sub); 
-            q_preprocessed[2] = (q_preprocessed[3]-2*q_preprocessed[0]+0)/eps_t_sub_2;
-            q_preprocessed[D_effective-2] = (0-q_preprocessed[D_effective-6])/(2*eps_t_sub); 
-            q_preprocessed[D_effective-1] = (0-2*q_preprocessed[D_effective-3]+q_preprocessed[D_effective-6])/eps_t_sub_2;
-            
+    COMPLEX (*r_from_q[3])(COMPLEX);
```
Comment (ShrinivasJC) [inline src/private/fnft__nse_discretization.c:399]: Please add a note here as to why three r_to_q functions are needed.

Comment (PJ-Prins) [reply src/private/fnft__nse_discretization.c:399]: That requirement is documented here:

https://github.com/FastNFT/FNFT/blob/972761817fdd892af1873b70b36b10457ef8ad5f/include/private/fnft__akns_discretization.h#L179

**Review by ShrinivasJC — APPROVED** [review]

File: src/private/fnft__nse_discretization.c (hunk 1 above)
Comment (wahls) [reply src/private/fnft__nse_discretization.c:399]: With the documentation in mind, why can we then choose `r_from_q[0]` == `r_from_q[1]` == `r_from_q[2]` in `fnft__nse_discretization_preprocess_signal`? Should they not be different?

**Comment by wahls** [comment]
Dear all,

could you please have a look at the recent commits? I a few more things I stumbled upon:

- `nse_fscatter` and `kdv_scatter` now do very little but add a layer of abstraction. Should we remove these functions and call `akns_fscatter` directly?

- Could the weights for CF6_4 in `akns_discretization_method_weights` be written in a way that is precise if one changes the precision of the COMPLEX type (as it was done e.g. for CF5_3)?

- The years in the headers of some of the changed files have not been updated.

**Comment by ShrinivasJC** [comment]
> * `nse_fscatter` and `kdv_scatter` now do very little but add a layer of abstraction. Should we remove these functions and call `akns_fscatter` directly?

I am not sure but I think our initial philosophy was that private functions with a specific prefix should only call functions with same prefix.

> * Could the weights for CF6_4 in `akns_discretization_method_weights` be written in a way that is precise if one changes the precision of the COMPLEX type (as it was done e.g. for CF5_3)?

Calculation of the weights of CF6_4 requires solving a system of nonlinear equations. Implementing that will be very involved without much gain at this stage.

**Comment by ShrinivasJC** [comment]
@wahls It shows that there is some conflict in the file src/fnft_nsev.c but for some reason I cannot see it. Can you please check it?

**Comment by wahls** [comment]
> > * `nse_fscatter` and `kdv_scatter` now do very little but add a layer of abstraction. Should we remove these functions and call `akns_fscatter` directly?
> I am not sure but I think our initial philosophy was that private functions with a specific prefix should only call functions with same prefix.

The `misc_` and `poly_` functions are frequently called from private functions with other prefixes. From that perspective it would be fine for me to also call `akns_` functions directly where appropriate.

> > * Could the weights for CF6_4 in `akns_discretization_method_weights` be written in a way that is precise if one changes the precision of the COMPLEX type (as it was done e.g. for CF5_3)?
> Calculation of the weights of CF6_4 requires solving a system of nonlinear equations. Implementing that will be very involved without much gain at this stage.

Any chance there is some Matlab source that we could put here in a comment in case they have to be re-computed?

**Comment by wahls** [comment]
> @wahls It shows that there is some conflict in the file src/fnft_nsev.c but for some reason I cannot see it. Can you please check it?

It should be fine now.

**Comment by ShrinivasJC** [comment]
> > @wahls It shows that there is some conflict in the file src/fnft_nsev.c but for some reason I cannot see it. Can you please check it?
> 
> It should be fine now.

Yes it is fixed now. Thank you.

**Comment by PJ-Prins** [comment]
> * `nse_fscatter` and `kdv_scatter` now do very little but add a layer of abstraction. Should we remove these functions and call `akns_fscatter` directly?

At least kdv_scatter should stay, to leave room for discretizations that are not in AKNS-basis.
(In principle the preprocessing of `q` is not specific to AKNS either, but the preprocessing of `r` is, and currently there is no better place to put it.)
I (still) think that the only way that the AKNS structure would be really advantageous when adding other PDE's (i.e. without code copying with some find and replace) is if `nsev.c` would currently merely be a wrapper of an `aknsv.c` with some NSE-specific function pointers. From `aknsv.c` you could call `akns_fscatter` and such.

> * The years in the headers of some of the changed files have not been updated.

Sorry, I will update it. Too bad github doesn't do that for us.

**Comment by ShrinivasJC** [comment]
> > > * `nse_fscatter` and `kdv_scatter` now do very little but add a layer of abstraction. Should we remove these functions and call `akns_fscatter` directly?
> > >   I am not sure but I think our initial philosophy was that private functions with a specific prefix should only call functions with same prefix.
> 
> The `misc_` and `poly_` functions are frequently called from private functions with other prefixes. From that perspective it would be fine for me to also call `akns_` functions directly where appropriate.

It is possible then. Just need to call the akns function to obtain the right akns discretization before calling akns_scatter.

> > > * Could the weights for CF6_4 in `akns_discretization_method_weights` be written in a way that is precise if one changes the precision of the COMPLEX type (as it was done e.g. for CF5_3)?
> > >   Calculation of the weights of CF6_4 requires solving a system of nonlinear equations. Implementing that will be very involved without much gain at this stage.
> 
> Any chance there is some Matlab source that we could put here in a comment in case they have to be re-computed?

I took the coefficients from the referenced paper. The authors haven't shared any code to compute them.

**Comment by wahls** [comment]
> At least kdv_scatter should stay, to leave room for discretizations that are not in AKNS-basis.

Ok, then let's keep them for now.

> I (still) think that the only way that the AKNS structure would be really advantageous when adding other PDE's (i.e. without code copying with some find and replace) is if `nsev.c` would currently merely be a wrapper of an `aknsv.c` with some NSE-specific function pointers. From `aknsv.c` you could call `akns_fscatter` and such.

The usefulness of this approach might also depend on how similar the structure of the spectra for other PDEs are. I'd propose to look into this when another PDE that fits AKNS is to be added.

> > * The years in the headers of some of the changed files have not been updated.
> Sorry, I will update it. Too bad github doesn't do that for us.

Thanks!

**Comment by wahls** [comment]
> It is possible then. Just need to call the akns function to obtain the right akns discretization before calling akns_scatter.

Let's keep it as it is for now (see other reply).
 
> Any chance there is some Matlab source that we could put here in a comment in case they have to be re-computed?
> 
> I took the coefficients from the referenced paper. The authors haven't shared any code to compute them.

Ok, bummer.

**Review by wahls — APPROVED** [review]
Looks good to me now. Unless there are other comments, I propose to merge.

**Comment by ShrinivasJC** [comment]
> Looks good to me now. Unless there are other comments, I propose to merge.

No more comments from me.

**Comment by PJ-Prins** [comment]
> > Looks good to me now. Unless there are other comments, I propose to merge.
> 
> No more comments from me.

Likewise.

**Comment by wahls** [comment]
Thanks, everyone!

**Pull request merged**